File: go-knowledge-app/internal/collectors/sessions_native.go
package collectors

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// Per-MTok USD rates (Claude Sonnet-4 tier) — kept identical to tokens.go.
const (
	rateInput      = 3.0
	rateOutput     = 15.0
	rateCacheWrite = 3.75
	rateCacheRead  = 0.30
)

// EstCost computes the estimated USD cost from token counts.
func EstCost(inp, out, cacheRead, cacheCreation float64) float64 {
	return (inp*rateInput + out*rateOutput + cacheCreation*rateCacheWrite +
		cacheRead*rateCacheRead) / 1_000_000
}

var nonAlnum = regexp.MustCompile(`[^A-Za-z0-9]`)

// EncodeCwd mirrors how Claude Code encodes a project's cwd into its session dir
// name (non-alnum → '-').
func EncodeCwd(cwd string) string {
	return nonAlnum.ReplaceAllString(cwd, "-")
}

func sessionsDir(cwd string) string {
	base := os.Getenv("CLAUDE_PROJECTS_DIR")
	if base == "" {
		base = filepath.Join(Home(), ".claude", "projects")
	}
	return filepath.Join(base, EncodeCwd(cwd))
}

type sessionUsage struct {
	InputTokens         float64 `json:"input_tokens"`
	OutputTokens        float64 `json:"output_tokens"`
	CacheReadTokens     float64 `json:"cache_read_input_tokens"`
	CacheCreationTokens float64 `json:"cache_creation_input_tokens"`
}

type sessionLine struct {
	Message *struct {
		Usage *sessionUsage `json:"usage"`
	} `json:"message"`
}

// FileCost returns the estimated USD cost from one session JSONL: sum usage
// tokens over its lines.
func FileCost(path string) float64 {
	b, err := os.ReadFile(path)
	if err != nil {
		return 0
	}
	var inp, out, cr, cc float64
	for _, line := range strings.Split(string(b), "\n") {
		if line == "" {
			continue
		}
		var e sessionLine
		if json.Unmarshal([]byte(line), &e) != nil {
			continue
		}
		if e.Message == nil || e.Message.Usage == nil {
			continue
		}
		u := e.Message.Usage
		inp += u.InputTokens
		out += u.OutputTokens
		cr += u.CacheReadTokens
		cc += u.CacheCreationTokens
	}
	return EstCost(inp, out, cr, cc)
}

// SevenDayCostNative sums the cost of session JSONLs whose file mtime is within
// the last 7 days. No subprocess.
func SevenDayCostNative(cwd string) float64 {
	return sevenDayCostNativeAt(cwd, time.Now().UnixMilli())
}

func sevenDayCostNativeAt(cwd string, nowMs int64) float64 {
	dir := sessionsDir(cwd)
	cutoff := nowMs - 7*24*3600*1000
	var cost float64
	entries, err := os.ReadDir(dir)
	if err != nil {
		return 0 // no session dir for this project
	}
	for _, ent := range entries {
		if ent.IsDir() || !strings.HasSuffix(ent.Name(), ".jsonl") {
			continue
		}
		p := filepath.Join(dir, ent.Name())
		info, err := os.Stat(p)
		if err != nil {
			continue
		}
		if info.ModTime().UnixMilli() < cutoff {
			continue
		}
		cost += FileCost(p)
	}
	return math.Round(cost*100) / 100
}

File: deno-knowledge-app/src/collectors/sessions_native.ts
// Native (Deno-only) session-cost reader — replaces the heavy `uv run
// session_analyze.py` call for the cross-project sidebar. Streams the same usage
// fields session_analyze.py sums (message.usage.{input,output,cache_read,
// cache_creation}_tokens) straight from the JSONL, so cost numbers stay in
// parity — but with no Python subprocess and tiny memory (one file at a time).

import { join } from "@std/path";

// Per-MTok USD rates (Claude Sonnet-4 tier) — kept identical to tokens.ts.
const RATE_INPUT = 3.0;
const RATE_OUTPUT = 15.0;
const RATE_CACHE_WRITE = 3.75;
const RATE_CACHE_READ = 0.30;

function estCost(
  inp: number,
  out: number,
  cacheRead: number,
  cacheCreation: number,
): number {
  return (inp * RATE_INPUT + out * RATE_OUTPUT + cacheCreation * RATE_CACHE_WRITE +
    cacheRead * RATE_CACHE_READ) / 1_000_000;
}

/** Claude Code encodes a project's cwd into its session dir name (non-alnum → '-'). */
function encodeCwd(cwd: string): string {
  return cwd.replace(/[^A-Za-z0-9]/g, "-");
}

function sessionsDir(cwd: string): string {
  const base = Deno.env.get("CLAUDE_PROJECTS_DIR") ??
    `${Deno.env.get("HOME") ?? "/tmp"}/.claude/projects`;
  return join(base, encodeCwd(cwd));
}

/** Estimated cost (≈ USD) from one session JSONL: sum usage tokens over its lines. */
async function fileCost(path: string): Promise<number> {
  let text: string;
  try {
    text = await Deno.readTextFile(path);
  } catch {
    return 0;
  }
  let inp = 0, out = 0, cr = 0, cc = 0;
  for (const line of text.split("\n")) {
    if (!line) continue;
    let e: { message?: { usage?: Record<string, unknown> } };
    try {
      e = JSON.parse(line);
    } catch {
      continue;
    }
    const usage = e?.message?.usage;
    if (!usage || typeof usage !== "object") continue;
    inp += Number(usage.input_tokens ?? 0);
    out += Number(usage.output_tokens ?? 0);
    cr += Number(usage.cache_read_input_tokens ?? 0);
    cc += Number(usage.cache_creation_input_tokens ?? 0);
  }
  return estCost(inp, out, cr, cc);
}

/**
 * Native 7-day cost (≈ USD): sum the cost of session JSONLs whose file mtime is
 * within the last 7 days. No subprocess. `now` is injectable for tests.
 */
async function sevenDayCostNative(cwd: string, now: number = Date.now()): Promise<number> {
  const dir = sessionsDir(cwd);
  const cutoff = now - 7 * 24 * 3600 * 1000;
  let cost = 0;
  try {
    for await (const ent of Deno.readDir(dir)) {
      if (!ent.isFile || !ent.name.endsWith(".jsonl")) continue;
      const p = join(dir, ent.name);
      let mt: number;
      try {
        mt = (await Deno.stat(p)).mtime?.getTime() ?? 0;
      } catch {
        continue;
      }
      if (mt < cutoff) continue;
      cost += await fileCost(p);
    }
  } catch { /* no session dir for this project */ }
  return Math.round(cost * 100) / 100;
}
